src/main/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTask.java

```diff
@@ -149,31 +194,72 @@ private void bulkWriteBatch(final List<MongoProcessedSinkRecordData> batch) {
     boolean bulkWriteOrdered = config.getBoolean(BULK_WRITE_ORDERED_CONFIG);
 
     Timer writeTime = Timer.start();
-    try {
-      LOGGER.debug(
-          "Bulk writing {} document(s) into collection [{}] via an {} bulk write",
-          writeModels.size(),
-          namespace.getFullName(),
-          bulkWriteOrdered ? "ordered" : "unordered");
-      BulkWriteResult result =
-          mongoClient
-              .getDatabase(namespace.getDatabaseName())
-              .getCollection(namespace.getCollectionName(), BsonDocument.class)
-              .bulkWrite(writeModels, new BulkWriteOptions().ordered(bulkWriteOrdered));
-      statistics.getBatchWritesSuccessful().sample(writeTime.getElapsedTime().toMillis());
-      statistics.getRecordsSuccessful().sample(batch.size());
-      LOGGER.debug("Mongodb bulk write result: {}", result);
-    } catch (RuntimeException e) {
-      statistics.getBatchWritesFailed().sample(writeTime.getElapsedTime().toMillis());
-      statistics.getRecordsFailed().sample(batch.size());
-      handleTolerableWriteException(
-          batch.stream()
-              .map(MongoProcessedSinkRecordData::getSinkRecord)
-              .collect(Collectors.toList()),
-          bulkWriteOrdered,
-          e,
-          config.logErrors(),
-          config.tolerateErrors());
+    int maxRetryCount = config.getRetryCount();
+    long retryIntervalMs = config.getRetryIntervalMs();
+    int retryCount = 0;
+    boolean shouldRetry = true;
+    while (shouldRetry) {
+      try {
+        LOGGER.debug(
+            "Bulk writing {} document(s) into collection [{}] via an {} bulk write",
+            writeModels.size(),
+            namespace.getFullName(),
+            bulkWriteOrdered ? "ordered" : "unordered");
+        BulkWriteResult result =
+            mongoClient
+                .getDatabase(namespace.getDatabaseName())
+                .getCollection(namespace.getCollectionName(), BsonDocument.class)
+                .bulkWrite(writeModels, new BulkWriteOptions().ordered(bulkWriteOrdered));
+        statistics.getBatchWritesSuccessful().sample(writeTime.getElapsedTime().toMillis());
+        statistics.getRecordsSuccessful().sample(batch.size());
+        LOGGER.debug("Mongodb bulk write result: {}", result);
+        break;
+      } catch (RuntimeException e) {
+        List<SinkRecord> sinkRecordList =
+            batch.stream()
+                .map(MongoProcessedSinkRecordData::getSinkRecord)
+                .collect(Collectors.toList());
+        if (config.tolerateDataErrors()) {
```

Comment:
Did you consider using Kafka connects own retry mechanism, which won't block the `put` call by throwining a [`RetryableException`](https://kafka.apache.org/25/javadoc/org/apache/kafka/common/errors/RetriableException.html). 

If not see [MongoSinkTask.java#L221-L257](https://github.com/mongodb/mongo-kafka/blob/r1.3.0/src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java#L221-L257) which demonstrates how the connector previously retried. This will reprocess the whole `put(final Collection<SinkRecord> records)` call. Which could be improved upon by storing the unprocessed records and just retrying them instead.

Reply:
I wasn't aware of this.  Thanks for making me aware of this method but I do think we need to do this in the connector itself because of the following.

I am expecting this retry to be transient.  I do see there is logic to split writes into multiple batches based on write model.   We could have got an exception during the processing of a middle batch.   If we throw the exception back to kafka, I think it would retry the entire group of batches.

However, in the case where we completely fail to connect to the sink, we will have this issue.